src/py21cmfast/wrapper/cfuncs.py

```diff
@@ -16,11 +17,82 @@
 
 logger = logging.getLogger(__name__)
 
+# Ideally, backend functions that we access here should do all the broadcasting/initialisation themselves
+# These decorators are for lower functions which are called directly in one or two lines, like delta_crit
+
+# TODO: a lot of these assume input as numpy arrays via use of .shape, explicitly require this
+
+
+def broadcast_params(func: Callable) -> Callable:
+    """Decorator to broadcast the parameters to the C library before calling the function."""
+
+    def wrapper(*args, **kwargs):
+        inputs = kwargs.get("inputs")
+        if inputs.astro_params:
+            lib.Broadcast_struct_global_all(
+                inputs.user_params.cstruct,
+                inputs.cosmo_params.cstruct,
+                inputs.astro_params.cstruct,
+                inputs.flag_options.cstruct,
+            )
+        else:
+            lib.Broadcast_struct_global_noastro(
+                inputs.user_params.cstruct,
+                inputs.cosmo_params.cstruct,
+            )
+        return func(*args, **kwargs)
+
+    return wrapper
+
+
+def init_backend_ps(func: Callable) -> Callable:
+    """Decorator to initialise the backend PS before calling the function."""
```

Comment:
Can we add some extra context to the docstring -- like that while this is not always necessary, it is very fast and means that we can confidently call any backend function knowing that the cosmo etc is up-to-date?